quickwit/quickwit-ingest/src/ingest_v2/mrecord.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright (C) 2023 Quickwit, Inc.
+//
+// Quickwit is offered under the AGPL v3.0 and as commercial software.
+// For commercial licensing, contact us at hello@quickwit.io.
+//
+// AGPL:
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, either version 3 of the
+// License, or (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+use bytes::{Buf, Bytes};
+use prost::Message;
+use quickwit_proto::ingest::{m_record_header, MRecordHeader};
+
+pub(super) enum MRecordPayload {
+    Doc(Bytes),
+    Commit,
+}
+
+pub(super) struct MRecord {
+    header: MRecordHeader,
+    payload: MRecordPayload,
+}
+
+impl MRecord {
+    pub fn new_doc(doc: Bytes) -> Self {
+        Self {
+            header: MRecordHeader::new_doc(),
+            payload: MRecordPayload::Doc(doc),
+        }
+    }
+
+    pub fn new_commit() -> Self {
+        Self {
+            header: MRecordHeader::new_commit(),
+            payload: MRecordPayload::Commit,
+        }
+    }
+
+    pub fn encode(&self) -> impl Buf {
+        // TODO: this is going to allocate for each record. No bueno.
+        let header_buf = Bytes::from(self.header.encode_length_delimited_to_vec());
```

Comment:
Here is the header allocation. Suggestions welcome :)

Reply:
The usual solution is to have a MRecordDocEncoder object that you add as a member of the `Ingester`. It holds an internal buffer and has a 
`fn encode(&mut self, mrecord: &MRecord) -> Buf`

Unfortunately in your case, lifetime will get in the way, because you want to use it in an iterator.

It is a none problem with the Iterator trait. It is also the raison d'etre of the Streamer trait for instance. https://docs.rs/fst/latest/fst/trait.Streamer.html
I don't see any simple solution unfortunately.